[Doc Improvement][Update TeamsFx-SDK][2753761] (#7405)
MicrosoftDocs/msteams-docs: msteams-platform/toolkit/TeamsFx-SDK.md

# TeamsFx SDK

> TeamsFx helps to reduce your tasks by using Microsoft Teams single sign-on (SSO) and accessing cloud resources down to single line statements with zero configuration. TeamsFx SDK can be used in browser and Node.js environments. TeamsFx core functionalities can be accessed in client and server environments. You can write user authentication code for:

Reviewer: ```suggestion
TeamsFx helps to reduce your tasks by using Microsoft Teams Single Sign-On (SSO) and accessing cloud resources down to single line statements with zero configuration. TeamsFx SDK can be used in browser and Node.js environments. TeamsFx core functionalities can be accessed in client and server environments. You can write user authentication code for:
```

The following code is an an example to create `TeamsUserCredential`:

```ts
const authConfig: TeamsUserCredentialAuthConfig = {
  clientId: process.env.REACT_APP_CLIENT_ID,
  initiateLoginEndpoint: process.env.REACT_APP_START_LOGIN_PAGE_URL,
};

const credential = new TeamsUserCredential(authConfig);
```

Required configurations are `initiateLoginEndpoint` and `clientId` which can be found inside type `TeamsUserCredentialAuthConfig`.

</details>

<details>
<summary><b> User identity in Node.js environment </b></summary>

`OnBehalfOfUserCredential` uses On-Behalf-Of flow and require Teams SSO token, in Azure Function or bot scenarios. TeamsFx SDK uses the following credential when you choose user identity in Node.js environment.

The following code is an example to create `OnBehalfOfUserCredential`:

```ts
const oboAuthConfig: OnBehalfOfCredentialAuthConfig = {
  authorityHost: process.env.M365_AUTHORITY_HOST,
  clientId: process.env.M365_CLIENT_ID,
  tenantId: process.env.M365_TENANT_ID,
  clientSecret: process.env.M365_CLIENT_SECRET,
};

const oboCredential = new OnBehalfOfUserCredential(ssoToken, oboAuthConfig);
```

> Required configurations are `authorityHost`, `tenantId`, `clientId`, `clientSecret` or `certificateContent` which can be found inside type `OnBehalfOfCredentialAuthConfig`.

Reviewer: ```suggestion
Required configurations are `authorityHost`, `tenantId`, `clientId`, `clientSecret`, or `certificateContent` which can be found inside type `OnBehalfOfCredentialAuthConfig`.
```

</details>

<details>
<summary><b> App identity in Node.js environment </b></summary>

> `AppCredential` represents the app identity. You can use app identity when user isn't involved, for example in a time-triggered automation job. TeamsFx SDK uses the following credential when you choose app identity in Node.js environment.

Reviewer: ```suggestion
`AppCredential` represents the app identity. You can use app identity when user isn't involved, for example, in a time-triggered automation job. TeamsFx SDK uses the following credential when you choose app identity in Node.js environment.
```

The following code is an example to create `AppCredential`:

```ts
const appAuthConfig: AppCredentialAuthConfig = {
  authorityHost: process.env.M365_AUTHORITY_HOST,
  clientId: process.env.M365_CLIENT_ID,
  tenantId: process.env.M365_TENANT_ID,
  clientSecret: process.env.M365_CLIENT_SECRET,
};
const appCredential = new AppCredential(appAuthConfig);
```

> Required configurations are `authorityHost`, `tenantId`, `clientId`, `clientSecret` or `certificateContent` which can be found inside type `AppCredentialAuthConfig`

Reviewer: ```suggestion
Required configurations are `authorityHost`, `tenantId`, `clientId`, `clientSecret`, or `certificateContent` which can be found inside type `AppCredentialAuthConfig`
```

## Microsoft Graph Scenarios

This section provides several code snippets for common scenarios that are related to the Microsoft Graph. In such scenarios, user can call APIs using different permissions in different ends (frontend/backend).

* User delegate permission in frontend (Use `TeamsUserCredential`)
    <details>
    <summary><b>Use graph API in tab app</b></summary>

    This code snippet shows you how to use `TeamsUserCredential` and `createMicrosoftGraphClientWithCredential` to get user profiles from Microsoft Graph in tab app. It also shows you how to catch and resolve a `GraphError`.

    1. Import the classes needed.

       ```typescript
       import {
        createMicrosoftGraphClientWithCredential,
        TeamsUserCredential,
       } from "@microsoft/teamsfx";
       ```

    2. Create `TeamsUserCredential` instance.

       ```typescript
       const authConfig: TeamsUserCredentialAuthConfig = {
       clientId: process.env.REACT_APP_CLIENT_ID,
       initiateLoginEndpoint: process.env.REACT_APP_START_LOGIN_PAGE_URL,
       };

       const teamsUserCredential = new TeamsUserCredential(authConfig);
       ```

>     2. Use `teamsUserCredential.login()` to get user consent.

Reviewer: Please update the number list order.

       ```typescript
       // Put these code in a call-to-action callback function to avoid browser blocking automatically showing up pop-ups.
       await teamsUserCredential.login(["User.Read"]); // Login with scope
       ```

>     3. You can initialize a TeamsFx instance and graph client and get information from MS Graph by this client.

Reviewer: update the number list order.

>     3. You can initialize a TeamsFx instance and graph client and get information from MS Graph by this client.

Reviewer: ```suggestion
1. You can initialize a TeamsFx instance and graph client and get information from Microsoft Graph by this client.
```

       const { DialogSet, WaterfallDialog } = require("botbuilder-dialogs");
       const { TeamsBotSsoPrompt, OnBehalfOfCredentialAuthConfig, TeamsBotSsoPromptSettings } = require("@microsoft/teamsfx");

       const convoState = new ConversationState(new MemoryStorage());
       const dialogState = convoState.createProperty("dialogState");
       const dialogs = new DialogSet(dialogState);

       const TeamsBotSsoPromptId = "TEAMS_BOT_SSO_PROMPT";

       const settings: TeamsBotSsoPromptSettings = {
       scopes: ["User.Read"],
       timeout: 900000,
       endOnInvalidMessage: true,
       };

       const authConfig: OnBehalfOfCredentialAuthConfig = {
        authorityHost: process.env.M365_AUTHORITY_HOST,
        clientId: process.env.M365_CLIENT_ID,
        tenantId: process.env.M365_TENANT_ID,
        clientSecret: process.env.M365_CLIENT_SECRET,
       };
       const loginUrl = process.env.INITIATE_LOGIN_ENDPOINT;
       const ssoPrompt = new TeamsBotSsoPrompt(authConfig, loginUrl, TeamsBotSsoPromptId, settings);

       dialogs.add(ssoPrompt);    
       ```

    2. Start the dialog and sign in.

>        ```typescript

Author: This code snippet is not well formatted, use the one below:

    ```typescript
    dialogs.add(
      new WaterfallDialog("taskNeedingLogin", [
        async (step) => {
          return await step.beginDialog("TeamsBotSsoPrompt");
        },
        async (step) => {
          const token = step.result;
          if (token) {
            // ... continue with task needing access token ...
          } else {
            await step.context.sendActivity(`Sorry... We couldn't log you in. Try again later.`);
            return await step.endDialog();
          }
        },
      ])
    );    
    ```

          return "Could not retrieve profile information from Microsoft Graph.";
        }
      }
     }    

     ```

>     For more information about how to use class in command bot, see [add single sign-on to Teams app](add-single-sign-on.md). And there's a [command-bot-with-sso](https://github.com/OfficeDev/TeamsFx-Samples/tree/dev/command-bot-with-sso) sample project, which you can try SSO command bot.

Author: For more information about how to use class in command bot
->
For more information about how to implement SSO command handler in command bot, ...

    </details>

    <details>
    <summary><b>Call Azure Function in tab app: On-Behalf-Of flow</b></summary>

    This code snippet shows you how to use `CreateApiClient` or `axios` library to call Azure Function, and how to call Graph API in Azure Function to get user profiles.

    1. You can use `CreateApiClient` provided by TeamsFx sdk to call Azure Function:

>        ```typescript

Author: The code is not well formatted, use the code below:

```typescript
async function callFunction() {
  const authConfig: TeamsUserCredentialAuthConfig = {
    clientId: process.env.REACT_APP_CLIENT_ID,
    initiateLoginEndpoint: process.env.REACT_APP_START_LOGIN_PAGE_URL,
  };
  const teamsUserCredential = new TeamsUserCredential(authConfig);
  // Create an API client by providing the token and endpoint.
  const apiClient = CreateApiClient(
    "https://YOUR_API_ENDPOINT", // Create an API Client that uses SSO token to authenticate requests
    new BearerTokenAuthProvider(async () =>  (await teamsUserCredential.getToken(""))!.token) // Call API hosted in Azure Functions on behalf of user to inject token to request header
  );
  // Send a GET request to "RELATIVE_API_PATH", "/api/functionName" for example.
  const response = await apiClient.get("RELATIVE_API_PATH");
  return response.data;
}    
```

       async function callFunction() {
        const authConfig: TeamsUserCredentialAuthConfig = {
        clientId: process.env.REACT_APP_CLIENT_ID,
        initiateLoginEndpoint: process.env.REACT_APP_START_LOGIN_PAGE_URL,
        };

       const teamsUserCredential = new TeamsUserCredential(authConfig);

        // Create an API client by providing the token and endpoint.
         const apiClient = CreateApiClient(
         "https://YOUR_API_ENDPOINT", // Create an API Client that uses SSO token to authenticate requests
         new BearerTokenAuthProvider(async () =>  (await teamsUserCredential.getToken(""))!.token) // Call API hosted in Azure Functions on behalf of user to inject token to request header
        );

        // Send a GET request to "RELATIVE_API_PATH", "/api/functionName" for example.
         const response = await apiClient.get("RELATIVE_API_PATH");
         return response.data;
       }    
       ```

        You can also use `axios` library to call Azure Function.

>         ```typescript

Author: Formated code as below: 

```typescript
async function callFunction() {
  const authConfig: TeamsUserCredentialAuthConfig = {
    clientId: process.env.REACT_APP_CLIENT_ID,
    initiateLoginEndpoint: process.env.REACT_APP_START_LOGIN_PAGE_URL,
  };
  const teamsUserCredential = new TeamsUserCredential(authConfig);
  const accessToken = await teamsUserCredential.getToken(""); // Get SSO token 
  const endpoint = "https://YOUR_API_ENDPOINT";
  const response = await axios.default.get(endpoint + "/api/" + functionName, {
    headers: {
      authorization: "Bearer " + accessToken.token,
    },
  });
  return response.data;
}    
```

    This code snippet shows you how to use certificate-based application permission to get the token that can be used to call Graph API.

    1. You can initialize the `appAuthConfig` by providing a `PEM-encoded key certificate`.

>        ```typescript

Author: ```typescript
 const appAuthConfig: AppCredentialAuthConfig = {
   authorityHost: process.env.M365_AUTHORITY_HOST,
   clientId: process.env.M365_CLIENT_ID,
   tenantId: process.env.M365_TENANT_ID,
   certificateContent: 'PEM-encoded key certificate',
 };
```

  1. Set the connection configuration.

     ```typescript
     // Equivalent to:
     // const sqlConnectConfig = new DefaultTediousConnectionConfiguration({
     //    sqlServerEndpoint: process.env.SQL_ENDPOINT,
     //    sqlUsername: process.env.SQL_USER_NAME,
     //    sqlPassword: process.env.SQL_PASSWORD,
     // });
     const teamsfx = new TeamsFx();
     // If there's only one SQL database
     const config = await getTediousConnectionConfig(teamsfx);
     // If there are multiple SQL databases
     const config2 = await getTediousConnectionConfig(teamsfx, "your database name");  
      ```

  2. Connect to your database.

     ```typescript
     const connection = new Connection(config);
     connection.on("connect", (error) => {
     if (error) {
      console.log(error);
      }
     });  
      ```

  > [!NOTE]

>   > For more information on sample to access SQL database in Azure Function, see [share-now sample](https://github.com/OfficeDev/TeamsFx-Samples/tree/dev/share-now). The

Author: For more information on sample to access SQL database in Azure Function, see [share-now sample](https://github.com/OfficeDev/TeamsFx-Samples/tree/dev/share-now). The
->
For more information on sample to access SQL database in Azure Function, see [share-now sample](https://github.com/OfficeDev/TeamsFx-Samples/tree/dev/share-now). ~~The~~

>   > `getTediousConnectionConfig` function has been deprecated, we recommend you to compose your own tedious configuration for better flexibility.

Author: The
  > `getTediousConnectionConfig` function ...

->
  > The `getTediousConnectionConfig` function

You can set customer log level and redirect outputs when using this library.

> [!NOTE]

> > Log in is turned off by default, you can turn it on by setting log level.

Author: Log in -> Logs ?

> > Sign in is turned off by default, you can turn it on by setting log level.

Author: We cannot use sign in here, it should be logging for printing logs
